### PR TITLE
Allow restricting login to specific countries

### PR DESCRIPTION
**Pull request by JoosuaKoskinen**
#### What are the main changes in this PR?

Adds GeoIP based login restricting. If the user tries to login from a country that's not in the list of countries from which login is allowed, error is shown on login (see screenshots).

Restrictions are only in use when there's at least on country on the list. So no restrictions enabled by default but if you were to add `Finland` to the list, you could only login from Finland from there on.

Adds WP-CLI commands to manage the restrictions:
```
PRODUCTION [@joosuakoskinen-4a5d38:~] $ wp seravo geologin
usage: wp seravo geologin allow <name> [--network]
   or: wp seravo geologin disallow <name> [--network]
   or: wp seravo geologin list 
```

The `geologin allow` adds a country to the allowed countries list. The `geologin disallow` removes entries from the allowed countries list. The `geologin list` tells whether the restrictions are enabled, what countries are allowed on the current blog and what countries are allowed on the whole multisite.

On multisites, the sites all have their own list and also the common network-wide list. For example below we have `geologin list` for a single blog "JoosuaKoskinen" in a multisite on which you could login from Finland, Norway and United Kingdom.
```
PRODUCTION [@joosuakoskinen-4a5d38:~] $ wp seravo geologin list
Restricted login: Enabled

Allow login on blog 'Joosuakoskinen' from:

    - Finland (FI)
    - Norway (NO)

Allow login network-wide (all blogs) from:

    - United Kingdom (UK)
```

#### Manual testing steps?

CLI needs to be tested on a regular and a multisite. Especially regular site as this was mainly developed on a multisite. The login blocking hasn't been tested much, I did a few tests with a VPN and seemed to work great.

#### Screenshot
geologin

### CODE REVIEW

**Review by sjaks — APPROVED** [review]
Tested with a VPN and works. Code looks good!

Just translations are lacking.

EDIT: I tested this on a non-multisite.

**Comment by JoosuaKoskinen** [comment]
> Just translations are lacking.

@sjaks do you think we should translate the CLI commands? In my opinion it might be bit weird to have CLI commands in finnish as `wp-*` commands are all only in english and the WP-CLI functions `WP_CLI::warning(...)`, `WP_CLI::error(...)` etc print english only.

**Comment by sjaks** [comment]
@JoosuaKoskinen good point :smile: I had in mind that at least the man/--help was translated but as it's not the case, this looks good as is.

**Review by simoke — COMMENTED** [review]
Tested with a test site:
```
PRODUCTION [@simo4:~] $ wp seravo geologin allow SE
Success: Login from 'Sweden' is now allowed.
PRODUCTION [@simo4:~] $ wp seravo geologin list
Restricted login: Enabled

Allow login from:

    - Sweden (SE)
```
... and got an error:
image
Login feature first checks the credentials, and then the origin country. If I type a wrong user name, the error message is the usual "Error: The username test is not registered on this site. If you are unsure of your username, try your email address instead."

Added FI to allow list and then login works normally:
```
PRODUCTION [@simo4:~] $ wp seravo geologin allow FI
Success: Login from 'Finland' is now allowed.
PRODUCTION [@simo4:~] $ wp seravo geologin list
Restricted login: Enabled

Allow login from:

    - Finland (FI)
    - Sweden (SE)
```
The log entry for the failed login is generic (as intended):
```
176.93.249.97 - seravo [15/Nov/2021:12:35:53 +0000] "POST /wp-login.php" 401 1000 "https://simo4.hel-test.seravo.dev/wp-login.php?redirect_to=https%3A%2F%2Fsimo4.hel-test.seravo.dev%2Fwp-admin%2F&reauth=1" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0" FAIL 
176.93.249.97 - testi [15/Nov/2021:12:37:58 +0000] "POST /wp-login.php" 401 1000 "https://simo4.hel-test.seravo.dev/wp-login.php" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0" FAIL 
```
Converted the site to multisite, and `wp seravo geologin list` output was changed:
```
PRODUCTION [@simo4:/data/wordpress/nginx] master(+11/-0,1)* ± wp seravo geologin list
Restricted login: Enabled

Allow login on blog 'Simo4' from:

    - Finland (FI)
    - Sweden (SE)

Allow login network-wide (all blogs) from:
```
I now created a subdir site, but should it possible to set rules to the individual sites now? At least the sites are not listed with `wp seravo geologin list` command:
```
PRODUCTION [@simo4:/data/wordpress/nginx] master(+11/-0,1)* ± wp site list
+---------+------------------------------------------------+---------------------+---------------------+
| blog_id | url                                            | last_updated        | registered          |
+---------+------------------------------------------------+---------------------+---------------------+
| 1       | https://simo4.hel-test.seravo.dev/             | 0000-00-00 00:00:00 | 2021-11-15 14:54:23 |
| 3       | https://simo4.hel-test.seravo.dev/subdir-test/ | 2021-11-15 12:57:37 | 2021-11-15 12:57:37 |
+---------+------------------------------------------------+---------------------+---------------------+
PRODUCTION [@simo4:/data/wordpress/nginx] master(+11/-0,1)* ± wp seravo geologin list
Restricted login: Enabled

Allow login on blog 'Simo4' from:

    - Finland (FI)
    - Sweden (SE)

Allow login network-wide (all blogs) from:


PRODUCTION [@simo4:/data/wordpress/nginx] master(+11/-0,1)* ±
```
Basically the feature seems to work really fine.

**Comment by JoosuaKoskinen** [comment]
> Login feature first checks the credentials, and then the origin country. If I type a wrong user name, the error message is the usual "Error: The username test is not registered on this site. If you are unsure of your username, try your email address instead."

Would it be better the other way?

> I now created a subdir site, but should it possible to set rules to the individual sites now? At least the sites are not listed with wp seravo geologin list command:

On multisites, you can specify the site to run the command on with eg. `wp seravo geologin list --url=https://simo4.hel-test.seravo.dev/subdir-test`. This applies to all WP-CLI commands, not only the ones from Seravo Plugin.

I could add an optional argument for `wp seravo geologin allow` to specify the subsite but could be confusing if there's two ways to specify the site (`--url` and the optional argument).

All subsites could be listed with `geologin list` but I think we have some sites with a huge list of subsites so the output would be awful.

**Comment by simoke** [comment]
> Would it be better the other way?
I suppose this is OK without changes :)

> On multisites, you can specify the site to run the command on with eg. wp seravo geologin list --url=https://simo4.hel-test.seravo.dev/subdir-test. This applies to all WP-CLI commands, not only the ones from Seravo Plugin.
>
>I could add an optional argument for wp seravo geologin allow to specify the subsite but could be confusing if there's two ways to specify the site (--url and the optional argument).
Tested this and works fine:
```
PRODUCTION [@simo4:/data/home/simo4] $ wp seravo geologin allow FI --url=https://simo4.hel-test.seravo.dev/subdir-test
Success: Login from 'Finland' is now allowed on blog 'Language test pahe'.
PRODUCTION [@simo4:/data/home/simo4] $ wp seravo geologin list --url=https://simo4.hel-test.seravo.dev/subdir-test
Restricted login: Enabled

Allow login on blog 'Language test pahe' from:

    - Finland (FI)

Allow login network-wide (all blogs) from:
```

**Review by simoke — APPROVED** [review]
See comments. LGTM!

**Comment by JoosuaKoskinen** [comment]
FYI @simoke 

> Login feature first checks the credentials, and then the origin country.

Just realized that of course the origin country should be checked first :grin: I made the following small changes:
- Removed `if ( is_wp_error($user) ) return $user;` from `verify_login_country` function and added `verify_login_country` function to `authenticate` filter. Now if the country is invalid, it overwrites other possible errors (like "invalid password" or "username not found") with the invalid country error.
- Prefixed the invalid country error with `<strong>Error</strong>: ` so it looks the same as other WP login errors.

As the changes are minor and tested myself afterwards, no more testing needed in my opinion. Will merge tomorrow unless gets change requests anyways.
